File: src/me/stormma/c/SortExample.c
# include <stdio.h>
# define SIZE 6

static int aux[SIZE] = {0};

void show_array(int* array) {
	for (int i = 0; i < SIZE; i++) {
        printf("%d ", array[i]);
    }
    printf("\n");
}

void exch(int* array, int a, int b) {
    int temp = array[b];
    array[b] = array[a];
    array[a] = temp;
}

void select_sort(int* array) {
	for (int i = 0; i < SIZE - 1; i++) {
		int min = i;
		for (int j = i + 1; j < SIZE; j++) {
			if (array[j] < array[min]) {
				min = j;
			}
		}
		if (min != i) {
			int temp = array[min];
			array[min] = array[i];
			array[i] = temp;
		}
	}
}

void bubble_sort(int* array) {
	for (int i = 0; i < SIZE - 1; i++) {
        for (int j = 0; j < SIZE - 1 - i; j++) {
            if (array[j] > array[j+1]) {
                int temp = array[j];
                array[j] = array[j+1];
                array[j+1] = temp;
            }
        }
    }
}

void insert_sort(int* array) {
	for (int i = 1; i < SIZE; i++) {
		for (int j = i; j > 0; j--) {
			if (array[j] < array[j-1]) {
				int temp = array[j];
				array[j] = array[j-1];
				array[j-1] = temp;
			}
		}
	}
}

void shell_sort(int* array, int len) {
	int gap = 1;
	while (gap < len / 3) {
		gap = gap * 3 + 1;
	}
	for (; gap > 0; gap /= 3) {
		for (int i = gap; i < len; i++) {
			for (int j = i; j >= gap; j -= gap) {
				if (array[j] < array[j-gap]) {
					int temp = array[j];
                	array[j] = array[j-gap];
                	array[j-gap] = temp;
				}
			}
		}
	}
}

void _merge(int* array, int low, int mid, int high) {
	int i = low;
	int j = mid + 1;
	for (int k = low; k <= high; k++) {
		aux[k] = array[k];
	}
	for (int k = low; k <= high; k++) {
		if (i > mid)
			array[k] = aux[j++];
		else if (j > high)
			array[k] = aux[i++];
		else if (aux[j] < aux[i]) {
			array[k] = aux[j++];
		} else {
			array[k] = aux[i++];
		}
	}
}

void _merge_sort(int* array, int low, int high) {
    if (high <= low)
        return;
    int mid = low + ((high - low) >> 1);
    _merge_sort(array, low, mid);
    _merge_sort(array, mid+1, high);
    _merge(array, low, mid, high);
}

void merge_sort(int * array, int length) {
    int high = length - 1, low = 0;
    _merge_sort(array, low, high);
}

int partition(int* array, int low, int high) {
    int i = low, j = high + 1;
    int value = array[low];
    while (1) {
        while (array[++i] < value) {
            if (i == high)
                break;
        }
        while (value < array[--j]) {
            if (j == low)
                break;
        }
        if (i >= j)
            break;
        exch(array, i, j);
    }
    exch(array, low, j);
    return j;
}

void _quick_sort(int* array, int low, int high) {
    if (high <= low) {
        return;
    }
    int j = partition(array, low, high);
    _quick_sort(array, low, j - 1);
    _quick_sort(array, j + 1, high);
}

void quick_sort(int* array, int length) {
    _quick_sort(array, 0, length - 1);
}

void _quick_sort_3way(int* array, int low, int high) {
    if (high <= low) {
        return;
    }
    int lt = low, i = low + 1, gt = high;
    int value = array[low];
    while (i <= gt) {
        int cmp = array[i] - value;
        if (cmp < 0)
            exch(array, lt++, i++);
        else if (cmp > 0)
            exch(array, i, gt--);
        else
            i++;
    }
    _quick_sort_3way(array, low, lt - 1);
    _quick_sort_3way(array, gt + 1, high);
}
void quick_sort_3way(int * array, int length) {
    _quick_sort(array, 0, length - 1);
}

int main() {
	int array[] = {3, 2, 1, 4, 9, 8};
	select_sort(array);
	printf("After select sort: \n");
	show_array(array);

	bubble_sort(array);
	printf("After bubble sort: \n");
	show_array(array);

	insert_sort(array);
	printf("After insert sort: \n");
	show_array(array);

	shell_sort(array, sizeof(array) / sizeof(array[0]));
	printf("After shell sort: \n");
	show_array(array);

	merge_sort(array, sizeof(array) / sizeof(array[0]));
	printf("After merge sort: \n");
	show_array(array);


    quick_sort(array, sizeof(array) / sizeof(array[0]));
    printf("After quick sort: \n");
    show_array(array);

    quick_sort_3way(array, sizeof(array) / sizeof(array[0]));
    printf("After quick three-way sort: \n");
    show_array(array);

    return 0;
}

File: src/me/stormma/java/SortExample.java
package me.stormma.java;

import java.util.Map;
import java.util.HashMap;

public class SortExample {

	public static void main(String[] args) {
//		Integer[] a = {1, 14, 2, 13, 3, 12, 4, 11, 5, 10, 6, 9, 7, 8};
		Integer[] a = {5, 1, 2, 4, 3, 6, 7, 9, 8};
		//Select Sort
		// me.stormma.java.SortUtils.sort(a, me.stormma.java.SortFactory.sortFactory.get("SelectSort"));
		// assert me.stormma.java.SortUtils.isSorted(a);
		// me.stormma.java.SortUtils.show(a);

		//Insert Sort
		// me.stormma.java.SortUtils.sort(a, me.stormma.java.SortFactory.sortFactory.get("InsertSort"));
		// assert me.stormma.java.SortUtils.isSorted(a);
		// me.stormma.java.SortUtils.show(a);

		//Bubble Sort
		// me.stormma.java.SortUtils.sort(a, me.stormma.java.SortFactory.sortFactory.get("BubbleSort"));
		// assert me.stormma.java.SortUtils.isSorted(a);
		// me.stormma.java.SortUtils.show(a);

		//Shell Sort
		// me.stormma.java.SortUtils.sort(a, me.stormma.java.SortFactory.sortFactory.get("ShellSort"));
		// assert me.stormma.java.SortUtils.isSorted(a);
		// me.stormma.java.SortUtils.show(a);

		//Top-Down Merge Sort
		//SortUtils.sort(a, SortFactory.sortFactory.get("MergeSortTD"));
		//assert SortUtils.isSorted(a);
		//SortUtils.show(a);

		//Bottom-Top Merge Sort
		//SortUtils.sort(a, SortFactory.sortFactory.get("MergeSortBU"));
		//assert SortUtils.isSorted(a);
		//SortUtils.show(a);

		SortUtils.sort(a, SortFactory.sortFactory.get("QuickSort"));
		assert SortUtils.isSorted(a);
		SortUtils.show(a);
	}
}

interface Sort {
	void sort(Comparable[] a);
}

class SortUtils {

	public static boolean less(Comparable v, Comparable w) {
		return v.compareTo(w) < 0;
	}

	public static void exch(Comparable[] a, int i, int j) {
		Comparable t = a[i];
		a[i] = a[j];
		a[j] = t;
		show(a);
	}

	public static void show(Comparable[] a) {
		for (int i = 0; i < a.length; i++) {
			System.out.print(a[i] + " ");
		}
		System.out.println();
	}

	public static boolean isSorted(Comparable[] a) {
		for (int i = 0; i < a.length; i++) {
			if (less(a[i], a[i-1])) {
				return false;
			}
		}
		return true;
	}

	public static void sort(Comparable[] a, Sort sort) {
		sort.sort(a);
	}
}

class SortFactory {
	public static Map<String, Sort> sortFactory;

	static {
		sortFactory = new HashMap<>();
		sortFactory.put("SelectSort", new SelectSort());
		sortFactory.put("BubbleSort", new BubbleSort());
		sortFactory.put("InsertSort", new InsertSort());
		sortFactory.put("ShellSort", new ShellSort());
		sortFactory.put("MergeSortTD", new MergeSortTD());
		sortFactory.put("MergeSortBU", new MergeSortBU());
		sortFactory.put("QuickSort", new QuickSort());
	}

	static class SelectSort implements Sort {
		@Override
		public void sort(Comparable[] a) {
			int length = a.length;
			for (int i = 0; i < length; i++) {
				int min = i;
				for (int j = i + 1; j < length; j++) {
					if (SortUtils.less(a[j], a[min])) {
						min = j;
					}
				}
				if (min != i) {
					SortUtils.exch(a, i, min);
				}
			}
		}
	}

	static class InsertSort implements Sort {
		@Override
		public void sort(Comparable[] a) {
			int length = a.length;
			for (int i = 1; i < length; i++) {
				for (int j = i; j > 0 && SortUtils.less(a[j], a[j-1]); j--) {
					SortUtils.exch(a, j, j-1);
				}
			}
		}
	}

	static class BubbleSort implements Sort {
		@Override
		public void sort(Comparable[] a) {
			int length = a.length;
			for (int i = 0; i < length - 1; i++) {
				for (int j = 0; j < length - 1 - i; j++) {
					if (!SortUtils.less(a[j], a[j+1])) {
						SortUtils.exch(a, j, j+1);
					}
				}
			}
		}
	}

	static class ShellSort implements Sort {
		@Override
		public void sort(Comparable[] a) {
			//before shell sort
			System.out.print("before shell sort==>");
			SortUtils.show(a);
			int length = a.length;
			int gap = 1; // 1, 4, 13, 40, 121...
			while (gap < length / 3) {
				gap = gap * 3 + 1;
			}
			for (; gap > 0; gap /= 3) {
				for (int i = gap; i < length; i++) {
					for (int j = i; j >= gap && SortUtils.less(a[j], a[j-gap]); j -= gap) {
						SortUtils.exch(a, j, j-gap);
					}
				}
			}
		}
	}

	static class MergeSortTD implements Sort {

		private static Comparable[] aux;

		@Override
		public void sort(Comparable[] a) {
			aux = new Comparable[a.length];
			sort(a, 0, a.length - 1);
		}

		private void sort(Comparable[] a, int low, int high) {
			if (high <= low) {
				return;
			}
			int mid = low + (high - low) / 2;
			sort(a, low, mid);
			sort(a, mid + 1, high);
			merge(a, low, mid, high);
		}

		private static void  merge(Comparable[] a, int low, int mid, int high) {
			int i = low, j = mid + 1;
			for (int k = low; k <= high; k++) {
				aux[k] = a[k];
			}
			for (int k = low; k <= high; k++) {
				if (i <= mid && j <= high) {
					a[k] = SortUtils.less(aux[i], aux[j]) ? aux[i++] : aux[j++];
				} else if (i > mid) {
					a[k] = aux[j++];
				} else if (j > high) {
					a[k] = aux[i++];
				}
			}
		}
	}

	static class MergeSortBU implements Sort {
		private static Comparable[] aux;
		@Override
		public void sort(Comparable[] a) {
			aux = new Comparable[a.length];
			int length = a.length;
			for (int sz = 1; sz < length; sz += sz) {
				for (int lo = 0; lo < length - sz; lo += 2 * sz) {
					System.out.println(sz);
					merge(a, lo, lo + sz - 1, Math.min(lo + 2 * sz - 1, length - 1) );
				}
			}
		}

		private static void  merge(Comparable[] a, int low, int mid, int high) {
			System.out.println("Starting => merge(a, " + low + ", " + mid + ", " + high + ")");
			int i = low, j = mid + 1;
			for (int k = low; k <= high; k++) {
				aux[k] = a[k];
			}
			for (int k = low; k <= high; k++) {
				if (i <= mid && j <= high) {
					a[k] = SortUtils.less(aux[i], aux[j]) ? aux[i++] : aux[j++];
				} else if (i > mid) {
					a[k] = aux[j++];
				} else if (j > high) {
					a[k] = aux[i++];
				}
			}
			System.out.println("Merge result: ");
			SortUtils.show(a);
		}
	}

	static class QuickSort implements Sort {

		@Override
		public void sort(Comparable[] a) {
			int length = a.length;
			sort(a, 0, length - 1);
		}

		private void sort(Comparable[] a, int low, int high) {
			if (high <= low) {
				return;
			}
			int j = partition(a, low, high);
			sort(a, low, j - 1);
			sort(a, j + 1, high);
		}

		private int partition(Comparable[] a, int low, int high) {
			int i = low, j = high + 1;
			Comparable value = a[low];
			while (true) {
				while (SortUtils.less(a[++i], value)) {
					if (i == high) {
						break;
					}
				}
				while (SortUtils.less(value, a[--j])) {
					if (j == low) {
						break;
					}
				}
				if (i >= j) {
					break;
				}
				SortUtils.exch(a, i, j);
			}
			SortUtils.exch(a, low, j);
			return j;
		}
	}
}
